src/core/qgsfeatureiterator.cpp

```diff
@@ -120,6 +121,37 @@ void QgsAbstractFeatureIterator::geometryToDestinationCrs( QgsFeature &feature,
   }
 }
 
+void QgsAbstractFeatureIterator::updateRequestToSourceCrs( QgsFeatureRequest &request, const QgsCoordinateTransform &transform ) const
+{
+  switch ( request.spatialFilterType() )
+  {
+    case Qgis::SpatialFilterType::NoFilter:
+      break;
+    case Qgis::SpatialFilterType::BoundingBox:
+    {
+      QgsRectangle newRect = transform.transformBoundingBox( request.filterRect(), Qgis::TransformDirection::Reverse );
+      request.setFilterRect( newRect );
+      break;
+    }
+    case Qgis::SpatialFilterType::DistanceWithin:
+    {
+      QgsGeometry geom = request.referenceGeometry();
+      QgsRectangle bbox = geom.boundingBox();
+      double x0 = bbox.xMaximum();
+      double y0 = bbox.yMaximum();
+      double dist = request.distanceWithin();
+      QgsPoint p0( x0, y0 );
+      QgsPoint p1( x0 + dist, y0 );
+      QgsLineString newDistLine( p0, p1 );
+      newDistLine.transform( transform, Qgis::TransformDirection::Reverse );
+      dist = newDistLine.length();
```

Comment:
I don't think this is a safe approach to do here. It's a good approach for an estimate, but we'll get differing results here depending on whether or not a transform is actually occurring. So it opens the possibility that users will get incorrect results.

I think instead what we need to do here is do the bounding box transform only (since that's always done as a worst-case scenario in transformBoundingBox), and then return false to indicate to the caller QgsVectorLayerFeatureIterator that the distance within check needs to be done locally (and accordingly we need to remove the limit from the provider's request).

Reply:
Sorry I don't fully understand the code, I just tried to improve what I was seeing already done.
What was done was setting mFilter to the reprojected bounding box and later re-setting the request bounding box to that rectangle:

     mRequest.setFilterRect( mFilterRect );

Maybe this was the problem: modifying the request in the first place ?

I'm not clear what doing things "locally" means (by QGIS instead of by provider?).

If you see what happens with the testcase included in this PR, my patch makes the filtering happen correctly on the provider's side (PostgreSQL case) and keeps the LIMIT.